### PR TITLE
Improve merge() algorithm

### PR DESCRIPTION
**Pull request by nordlow**
This is revival of my previous https://github.com/dlang/phobos/pull/3315 applied on top of existing implementation of `merge` that sneaked in without me noticing it ;)

**Improvements**:
- Bidirectional support if all inputs are bidirectional
- Propagates infiniteness if any of the inputs is infinite.
- Add in-place checking of sortedness in `popFront` and `popBack`. Done by comparing current and next element if element type allows it.

**Other notes**:
- Checking for `isSortedRange` in my previous implementation has been removed because we currently cannot compare all the different formats for the comparison alias (`pred`) fed to `SortedRange`.

**Update**:
I just noticed that `Merge` is used in setops. I guess this removes the opportunity for making this a Voldermort type, right?

### CODE REVIEW

File: std/algorithm/sorting.d [hunk 1]
```diff
@@ -1048,147 +1048,277 @@ unittest
...
+    {
+        private size_t _lastBackIndex = size_t.max;
+    }
 }
 
-/// Ditto
-Merge!(less, Rs) merge(alias less = "a < b", Rs...)
-(Rs rs)
+/**
+   Merge multiple sorted ranges $(D rs) with less-than predicate function $(D
```
Comment (andralex) [inline std/algorithm/sorting.d]: for single words prefer backticks to the D macro

File: std/algorithm/sorting.d [hunk 2]
```diff
@@ -1048,147 +1048,277 @@ unittest
...
+   pred) into one single sorted output range containing the sorted union of the
+   elements of inputs. Elements in the output are not unique, that is the length
```
Comment (andralex) [inline std/algorithm/sorting.d]: s/that is/consequently/

Better yet, rephrase the sentence: "Duplicates are not eliminated, meaning that the total number of elements in the result of `merge` is the sum of all elements in the ranges passed to it."

File: std/algorithm/sorting.d [hunk 3]
```diff
@@ -1048,147 +1048,277 @@ unittest
...
+Returns:
+    A range containing the union of the given ranges.
+
+See_Also:
+   $(XREF algorithm, setops, SetUnion)
+
+Details:
+
+All of its inputs are assumed to be sorted. This can mean that inputs are
+   instances of $(XREF range, SortedRange). Use the result of $(XREF algorithm,
```
Comment (andralex) [inline std/algorithm/sorting.d]: there is no such restriction in `Merge` or `merge`

File: std/algorithm/sorting.d [hunk 4]
```diff
@@ -1048,147 +1048,277 @@ unittest
...
+   ensuring that two or more instances of $(XREF range, SortedRange) are sorted
+   using a specific comparison function $(D pred). Therefore no checking is done
+   here to assure that the $(D pred) template parmeter matches the $(D pred)
+   template parameter that were given to each instance of $(XREF range,
+   SortedRange) when the $(D rs) were instantiated.
+
+   This algorithm is lazy, doing work progressively as elements are pulled off
+   the result.
+
+   Average complexity is $(BIGOH n * k) for $(D k) inputs of maximum length $(D n).
```
Comment (andralex) [inline std/algorithm/sorting.d]: "Time complexity is proportional to the sum of element counts over all ranges."

File: std/algorithm/sorting.d [hunk 5]
```diff
@@ -1048,147 +1048,277 @@ unittest
...
+   Example:
+   -------
+   auto a = [0, 2, 4];
+   auto b = [1, 3, 5];
+   auto c = merge(a.assumeSorted, b.assumeSorted)
+   assert(equal(c, [0, 1, 2, 3, 4, 5]));
+   assert(equal(c.retro, [5, 4, 3, 2, 1, 0]));
+   -------
+*/
+Merge!(less, Rs) merge(alias less = "a < b", Rs...)(Rs rs)
```
Comment (andralex) [inline std/algorithm/sorting.d]: The constraints should go here, in the user-visible and documented API, not in the `Merge` struct. The constraint should not use the package-level `CommonElementType`.

File: std/algorithm/sorting.d [hunk 6]
```diff
@@ -1048,147 +1048,277 @@ unittest
...
+    assert(equal(m, [10, 20, 30]));
+    m.popFront();
+    assert(equal(m, [20, 30]));
+    m.popBack();
+    assert(equal(m, [20]));
+    m.popFront();
+    assert(m.empty);
+}
+
+package template CommonElementType(Rs...)
```
Comment (andralex) [inline std/algorithm/sorting.d]: this is superfluous

**Comment by andralex** [comment]
You may assert that the ranges are sorted opportunistically; an eager call to isSorted seems a bit much.

**Comment by nordlow** [comment]
So no debug assertions of sortedness of output either, then?

File: std/algorithm/sorting.d (hunk 1 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: Done.

File: std/algorithm/sorting.d (hunk 3 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: There is on `Merge`. Add restriction on `merge`.

File: std/algorithm/sorting.d (hunk 4 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: Done.

File: std/algorithm/sorting.d (hunk 5 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: Should I remove restriction on `Merge`, then?

File: std/algorithm/sorting.d (hunk 6 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: Done.

File: std/algorithm/sorting.d (hunk 3 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: Ahh, hang on I'll remove this doc.

**Comment by nordlow** [comment]
Only thing left now is potential checking for sortedness in `frontIndex`.

File: std/algorithm/sorting.d [hunk 7]
```diff
@@ -1048,147 +1048,275 @@ unittest
...
+
+Params:
+    less = Predicate the given ranges are sorted by.
+    rs = The ranges to compute the union for.
+
+Returns:
+    A range containing the union of the given ranges.
+
+See_Also:
+   $(XREF algorithm, setops, SetUnion)
```
Comment (aG0aep6G) [inline std/algorithm/sorting.d]: Please use REF instead of XREF throughout.

Comment (nordlow) [inline std/algorithm/sorting.d]: Just here or everywhere?

Comment (nordlow) [inline std/algorithm/sorting.d]: Done.

File: std/algorithm/sorting.d [hunk 8]
```diff
@@ -1048,147 +1048,275 @@ unittest
...
+
+Params:
+    less = Predicate the given ranges are sorted by.
+    rs = The ranges to compute the union for.
+
+Returns:
+    A range containing the union of the given ranges.
+
+See_Also:
+   $(REF algorithm, setops, SetUnion)
```
Comment (aG0aep6G) [inline std/algorithm/sorting.d]: REF has a different order of arguments than XREF. Symbol name goes first, followed by the module including std. Here: `REF(SetUnion, std,algorithm,setops)`. However, there doesn't seem to be a std.algorithm.setops.SetUnion.

Comment (nordlow) [inline std/algorithm/sorting.d]: setUnion has been deprecated.

Comment (nordlow) [inline std/algorithm/sorting.d]: Done.

File: std/algorithm/sorting.d [hunk 9]
```diff
@@ -1048,147 +1048,275 @@ unittest
...
+
+Params:
+    less = Predicate the given ranges are sorted by.
+    rs = The ranges to compute the union for.
+
+Returns:
+    A range containing the union of the given ranges.
+
+See_Also:
+   $(REF setUnion, std,algorithm,setops)
```
Comment (aG0aep6G) [inline std/algorithm/sorting.d]: Alright, syntax is correct now. But the link is dead because setUnion has been undocumented. I suggest to just drop the See_Also section.

Comment (nordlow) [inline std/algorithm/sorting.d]: Done.

**Comment by aG0aep6G** [comment]
Ok, REFs look good to me.

**Comment by nordlow** [comment]
Added debug-only efficient checking for sortedness inside `Merge.pop{Front,Back}`.

**Comment by nordlow** [comment]
@andralex ping

**Comment by andralex** [comment]
As I mentioned via email I'm on the road this week.

**Comment by nordlow** [comment]
Ok. Thanks, @andralex 

Can anyone else approve this?

File: std/algorithm/sorting.d [hunk 10]
```diff
@@ -1048,147 +1048,287 @@ unittest
...
+
+   Example:
+   -------
+   auto a = [0, 2, 4];
+   auto b = [1, 3, 5];
+   auto c = merge(a.assumeSorted, b.assumeSorted)
+   assert(equal(c, [0, 1, 2, 3, 4, 5]));
+   assert(equal(c.retro, [5, 4, 3, 2, 1, 0]));
+   -------
+*/
```
Comment (CyberShadow) [inline std/algorithm/sorting.d]: Unless I'm missing something, the documentation should have a `See_also` section linking to `setUnion`, and an explanation for how this function is different from it.

Comment (CyberShadow) [inline std/algorithm/sorting.d]: Oh, that's because it's been deprecated in favor of `merge`. I guess I was missing something after all. Never mind :)

**Comment by nordlow** [comment]
So I guess I should remove the ref again then?

**Comment by CyberShadow** [comment]
Yes, we shouldn't link to deprecated symbols. Sorry.

**Comment by nordlow** [comment]
Restore.

**Comment by nordlow** [comment]
ping, @andralex

File: std/algorithm/sorting.d [hunk 11]
```diff
@@ -1048,147 +1048,287 @@ unittest
...
+    if (Rs.length >= 2 &&
+        allSatisfy!(isInputRange, Rs) &&
+        is(CommonType!(staticMap!(ElementType, Rs))))
 {
-private:
-    Rs _r;
-    alias comp = binaryFun!(less);
-    uint _crt;
+    alias ElementType = CommonType!(staticMap!(.ElementType, Rs));
+    enum isBidirectional = allSatisfy!(isBidirectionalRange, staticMap!(Unqual, Rs));
```
Comment (andralex) [inline std/algorithm/sorting.d]: make private

File: std/algorithm/sorting.d [hunk 12]
```diff
@@ -1048,147 +1048,287 @@ unittest
...
+    this(Rs source)
+    {
+        this.source = source;
+        this._lastFrontIndex = frontIndex;
+        static if (isBidirectional)
+            this._lastBackIndex = backIndex;
```
Comment (andralex) [inline std/algorithm/sorting.d]: This may be problematic. A vast fraction of `Merge` uses is in an input/forward manner, even though the inputs are bidirectional. Therefore computing eagerly matters related to backwards iteration is often adding overhead. Can you arrange computation such that any work related to backward iteration is done lazily?

File: std/algorithm/sorting.d [hunk 13]
```diff
@@ -1048,147 +1048,287 @@ unittest
...
 
-    void adjustPosition(uint candidate = 0)()
+    import std.functional : binaryFun;
+    private alias comp = binaryFun!less;
+
+    import std.typetuple : anySatisfy;
+    static if (anySatisfy!(isInfinite, Rs))
     {
-        static if (candidate == Rs.length)
+        enum bool empty = false; // propagate infiniteness
```
Comment (andralex) [inline std/algorithm/sorting.d]: nice

File: std/algorithm/sorting.d [hunk 14]
```diff
@@ -1048,147 +1048,287 @@ unittest
...
             {
-                adjustPosition!(candidate + 1)();
-                return;
+            case i:
+                assert(!source[i].empty);
+                return source[i].front;
             }
-            foreach (i, U; Rs[candidate + 1 .. $])
+        }
+        assert(false);
```
Comment (andralex) [inline std/algorithm/sorting.d]: I assume you don't need this because the switch is `final`.

File: std/algorithm/sorting.d [hunk 15]
```diff
@@ -1048,147 +1048,287 @@ unittest
...
+    }
+
+    private size_t frontIndex()
+    {
+        size_t bestIndex = size_t.max; // indicate undefined
+        Unqual!ElementType bestElement;
+        foreach (i, _; Rs)
+        {
+            if (!source[i].empty)
```
Comment (andralex) [inline std/algorithm/sorting.d]: `if (!source[i].empty) continue;`

Then the bulky logic is the loop body.

File: std/algorithm/sorting.d [hunk 16]
```diff
@@ -1048,147 +1048,287 @@ unittest
...
+        final switch (_lastFrontIndex)
+        {
+            foreach (i, _; Rs)
+            {
+            case i:
+                debug Unqual!ElementType previousFront = source[i].front();
+                source[i].popFront();
+                debug if (!source[i].empty)
+                {
+                    assert(previousFront == source[i].front ||
```
Comment (andralex) [inline std/algorithm/sorting.d]: This doesn't work if you have an input range, e.g. byLine. I suggest just remove this checking code.

File: std/algorithm/sorting.d [hunk 17]
```diff
@@ -1048,147 +1048,287 @@ unittest
...
+                           comp(previousFront, source[i].front),
+                           "Input " ~ i.stringof ~ " is unsorted"); // @nogc
+                }
+                break;
```
Comment (andralex) [inline std/algorithm/sorting.d]: This breaks the foreach loop but not the switch. So you'd need a break for the switch as well.

File: std/algorithm/sorting.d [hunk 18]
```diff
@@ -1048,147 +1048,287 @@ unittest
...
 
-    ///
-    @property bool empty()
+    static if (isBidirectional)
     {
-        return _crt == _crt.max;
-    }
+        @property auto ref back()
+        {
+            final switch (_lastBackIndex)
```
Comment (andralex) [inline std/algorithm/sorting.d]: So here _lastBackIndex would be initialized lazily.

File: std/algorithm/sorting.d [hunk 19]
```diff
@@ -1048,147 +1048,287 @@ unittest
...
-    }
 
-    ///
-    @property auto ref ElementType front()
-    {
-        assert(!empty);
-        // Assume _crt is correct
-        foreach (i, U; Rs)
+        void popBack()
         {
```
Comment (andralex) [inline std/algorithm/sorting.d]: Here, too, _lastBackIndex would be initialized lazily.

File: std/algorithm/sorting.d [hunk 20]
```diff
@@ -1048,147 +1048,287 @@ unittest
...
+                case i:
+                    debug Unqual!ElementType previousBack = source[i].back();
+                    source[i].popBack();
+                    debug if (!source[i].empty)
+                    {
+                        assert(previousBack == source[i].back ||
+                               comp(source[i].front, previousBack),
+                               "Input " ~ i.stringof ~ " is unsorted"); // @nogc
+                    }
+                    break;
```
Comment (andralex) [inline std/algorithm/sorting.d]: Same discussion about break. I'm really not sure what happens here because there's no code to fall through to.

File: std/algorithm/sorting.d [hunk 21]
```diff
@@ -1048,147 +1048,287 @@ unittest
...
+   auto b = [1, 3, 5];
+   auto c = merge(a.assumeSorted, b.assumeSorted)
+   assert(equal(c, [0, 1, 2, 3, 4, 5]));
+   assert(equal(c.retro, [5, 4, 3, 2, 1, 0]));
+   -------
+*/
+Merge!(less, Rs) merge(alias less = "a < b", Rs...)(Rs rs)
+    if (Rs.length >= 2 &&
+        allSatisfy!(isInputRange, Rs) &&
+        is(CommonType!(staticMap!(ElementType, Rs))))
```
Comment (andralex) [inline std/algorithm/sorting.d]: See comment above for the constraint (I recall `CommonType` when there's no actual common type is `void`).

File: std/algorithm/sorting.d [hunk 22]
```diff
@@ -1048,147 +1048,287 @@ unittest
...
-    A range containing the union of the given ranges.
-
-See_Also:
-   $(XREF algorithm, setops, SetUnion)
- */
-struct Merge(alias less = "a < b", Rs...) if (allSatisfy!(isInputRange, Rs))
+struct Merge(alias less = "a < b", Rs...)
+    if (Rs.length >= 2 &&
+        allSatisfy!(isInputRange, Rs) &&
+        is(CommonType!(staticMap!(ElementType, Rs))))
```
Comment (nordlow) [inline std/algorithm/sorting.d]: Changed to you version.

File: std/algorithm/sorting.d (hunk 11 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: Done.

File: std/algorithm/sorting.d (hunk 14 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: Fixed.

File: std/algorithm/sorting.d (hunk 21 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: Done.

File: std/algorithm/sorting.d (hunk 12 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: Further, what do you say we make counters `_last{Front,Back}Index` on uint on 64-bit and ushort on 32-bit to save one word of memory usage?

File: std/algorithm/sorting.d (hunk 15 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: Done.

File: std/algorithm/sorting.d (hunk 16 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: Do you mean in the case where `source[i]` is a, so called, volatile range?

File: std/algorithm/sorting.d (hunk 17 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: How do I accomplish that?

File: std/algorithm/sorting.d (hunk 18 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: Done.

File: std/algorithm/sorting.d (hunk 19 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: Got it. Done.

File: std/algorithm/sorting.d (hunk 20 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: These `foreach`es are static ones. I'm pretty sure that the break associates with `switch` here. If it we're I don't see how the unittests could pass.

**Comment by aG0aep6G** [comment]
On 11/21/2016 10:53 PM, Per Nordlöw wrote:

> These |foreach|es are static ones. I'm pretty sure that the break 
> associates with |switch| here. If it we're I don't see how the 
> unittests could pass.

You're mistaken. The break applies to the foreach. The behavior is a bit 
weird, though: The break doesn't end the static foreach loop. It jumps 
behind the foreach body at run-time, still inside the switch. If there's 
code there (e.g. a default case) that gets executed. There may be other 
subtle differences.

To break the switch instead, use a label:

```d
sw: switch (foo)
{
     foreach (x; alias_seq)
     {
         case x: /* ... */ break sw;
     }
}
```

**Comment by nordlow** [comment]
Ok, @aG0aep6G I'll fix.

**Comment by nordlow** [comment]
@aG0aep6G Changed the two switch breaks to be labeled.

**Comment by nordlow** [comment]
And cleaned up the order of the imports and private defs at the top of `Merge`.

**Comment by aG0aep6G** [comment]
> @aG0aep6G, what do you think about the updated limiting of the sortedness logic?

Sorry, I'm not following the actual objective of this PR. Just skimming the messages.

By the way, I think you've edited in that question later on. It's not in the notification email from GItHub. I wouldn't have noticed it if I hadn't had this page open in the browser.

Also, you seem to update the PR by amending. In my experience, that makes it much harder to review the changes. I'd suggest to squash once at the end instead of amending. But Andrei does the actual review here, so that's his call.

**Comment by nordlow** [comment]
Removed `isSorted` calls in unittest.

**Comment by nordlow** [comment]
Ok, I'll change to not amending and then squashing at the end.

File: std/algorithm/sorting.d [hunk 23]
```diff
@@ -1048,147 +1048,294 @@ unittest
...
-    void adjustPosition(uint candidate = 0)()
+    this(Rs source)
     {
-        static if (candidate == Rs.length)
+        this.source = source;
+        this._lastFrontIndex = frontIndex;
+        static if (isBidirectional)
         {
-            _crt = _crt.max;
+            this._lastBackIndex = size_t.max; // uinitialized, lazy initialized in `front`
```
Comment (andralex) [inline std/algorithm/sorting.d]: Why is this here and not in the initializer of the field?

File: std/algorithm/sorting.d [hunk 24]
```diff
@@ -1048,147 +1048,294 @@ unittest
...
-    ///
-    @property bool empty()
+    private size_t frontIndex()
     {
-        return _crt == _crt.max;
+        size_t bestIndex = size_t.max; // indicate undefined
+        Unqual!ElementType bestElement;
+        foreach (i, _; Rs)
+        {
+            if (source[i].empty) { continue; }
```
Comment (andralex) [inline std/algorithm/sorting.d]: This is unusual, for single statements on one line we omit the braces (except when required e.g. for function definitions).

File: std/algorithm/sorting.d [hunk 25]
```diff
@@ -1048,147 +1048,294 @@ unittest
...
+                        debug ElementType previousBack = source[i].back();
+                    source[i].popBack();
+                    static if (isCopyable!ElementType && !hasAliasing!ElementType)
+                        debug if (!source[i].empty)
+                        {
+                            assert(previousBack == source[i].back ||
+                                   comp(source[i].front, previousBack),
+                                   "Input " ~ i.stringof ~ " is unsorted"); // @nogc
+                        }
+                    break sw;
```
Comment (andralex) [inline std/algorithm/sorting.d]: as an aside, the label is not necessary - you may just leave the break here and insert one right after the foreach loop

File: std/algorithm/sorting.d [hunk 26]
```diff
@@ -1048,147 +1048,294 @@ unittest
...
+                result += source[i].length;
             }
             return result;
         }
 
-        ///
         alias opDollar = length;
     }
+
+    public Rs source;
```
Comment (andralex) [inline std/algorithm/sorting.d]: we usually put the state at the beginning

File: std/algorithm/sorting.d (hunk 23 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: Removed this assignments because it's initialized to `size_t` in declaration.

File: std/algorithm/sorting.d (hunk 24 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: Done.

File: std/algorithm/sorting.d (hunk 25 above)
Comment (andralex) [inline std/algorithm/sorting.d]: The suggestion is to have two unlabeled breaks, one inside the foreach and another one right after it. But it's really up to you.

**Comment by nordlow** [comment]
Rebased on master. One thing has changed: improved sortedness checking by breaking out CT-logic in `canCheckSortedness`.

**Comment by nordlow** [comment]
ci/circleci — Your tests failed on CircleCI fails. Don't know why.

File: std/algorithm/sorting.d [hunk 27]
```diff
@@ -1048,144 +1048,300 @@ unittest
...
+   `false`).
+
+   Example:
+   -------
+   auto a = [0, 2, 4];
+   auto b = [1, 3, 5];
+   auto c = merge(a.assumeSorted, b.assumeSorted)
+   assert(equal(c, [0, 1, 2, 3, 4, 5]));
+   assert(equal(c.retro, [5, 4, 3, 2, 1, 0]));
+   -------
```
Comment (wilzbach) [inline std/algorithm/sorting.d]: Any reason why this is a Ddoc unittest block?
If you put it as a normal `unittest` block and annotate it with the magic `///` slashes it will be displayed in the same way, but it has the advantage that we can ensure that the examples does compile. Another advantage is that soon the users will be able to directly run unittests blocks on dlang.org and then of course we will enable this only for tests for which we can ensure they compile & work.

File: std/algorithm/sorting.d [hunk 28]
```diff
@@ -1048,144 +1048,300 @@ unittest
...
+        allSatisfy!(isInputRange, Rs) &&
+        !is(CommonType!(staticMap!(ElementType, Rs)) == void))
 {
     return typeof(return)(rs);
 }
 
+/// test bi-directional access and common type
+@safe pure nothrow unittest
+{
+    import std.algorithm : equal;
```
Comment (wilzbach) [inline std/algorithm/sorting.d]: Output of `make -f posix.mak style`:

```
grep -nr 'import std.algorithm : ' $(find . -name '*.d') ; test $? -eq 1
./std/algorithm/sorting.d:1308:    import std.algorithm : equal;
make: *** [style] Error 1
```

-> `std.algorithm.comparison`

File: std/algorithm/sorting.d [hunk 29]
```diff
@@ -1048,144 +1048,300 @@ unittest
...
+    alias I = int;
+    alias D = double;
+
+    S[] a = [1, 2, 3];
+    I[] b = [50, 60];
+    D[] c = [10, 20, 30, 40];
+
+    auto m = merge(a, b, c);
+
+    static assert(is(typeof(m.front) == CommonType!(S, I, D)));
```
Comment (wilzbach) [inline std/algorithm/sorting.d]: output of `make -f posix.mak publictests`:

```
executing out/std_algorithm_sorting.d
out/std_algorithm_sorting.d(165): Error: template instance CommonType!(S, I, D) template 'CommonType' is not defined
out/std_algorithm_sorting.d(165): Error: static assert  (is(double == _error_)) is false
make: *** [publictests] Error 1
```

As mentioned above we want to ensure that public tests do compile without missing imports and hence we started to automatically check for it. So a public `unittest` block can't use the global imports.
Solution 1: use the ddoced example as public unittest and make this one private
Solution 2: add the missing import -> `import std.traits : CommonType`

File: std/algorithm/sorting.d [hunk 30]
```diff
@@ -1048,144 +1048,300 @@ unittest
...
     {
-        this._r = rs;
-        adjustPosition();
-    }
-
-    ///
-    @property bool empty()
-    {
-        return _crt == _crt.max;
+        size_t bestIndex = size_t.max; // indicate undefined
```
Comment (wilzbach) [inline std/algorithm/sorting.d]: With `ptrdiff_t` you could use `-1` - this seems to be used at a couple of similar bits at Phobos, e.g.
https://github.com/dlang/phobos/blob/master/std/algorithm/searching.d#L881
https://github.com/dlang/phobos/blob/master/std/string.d#L383

Explanation:
http://en.cppreference.com/w/cpp/types/ptrdiff_t

Comment (nordlow) [inline std/algorithm/sorting.d]: Why is `ptrdiff_t` preferred over `size_t` here?

File: std/algorithm/sorting.d (hunk 27 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: I didn't add this. It's ok with move to move it to a unittest if ok with the original author.

File: std/algorithm/sorting.d (hunk 28 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: Done

File: std/algorithm/sorting.d [hunk 31]
```diff
@@ -1048,152 +1048,302 @@ unittest
...
 ///
 @safe pure nothrow unittest
 {
     import std.algorithm.comparison : equal;
 
     int[] a = [1, 3, 5];
     int[] b = [2, 3, 4];
+
     assert(a.merge(b).equal([1, 2, 3, 3, 4, 5]));
+    assert(a.merge(b).retro.equal([5, 4, 3, 3, 2, 1]));
```
Comment (wilzbach) [inline std/algorithm/sorting.d]: `make -f publictests` or [CircleCi](https://circleci.com/gh/dlang/phobos/1035?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)

```
out/std_algorithm_sorting.d(202): Error: no property 'retro' for type 'Merge!("a < b", int[], int[])'
```

File: std/algorithm/sorting.d [hunk 32]
```diff
@@ -1048,152 +1048,302 @@ unittest
...
+    S[] a = [1, 2, 3];
+    I[] b = [50, 60];
+    D[] c = [10, 20, 30, 40];
+
+    auto m = merge(a, b, c);
+
+    static assert(is(typeof(m.front) == CommonType!(S, I, D)));
+
+    assert(equal(m, [1, 2, 3, 10, 20, 30, 40, 50, 60]));
+    assert(equal(m.retro, [60, 50, 40, 30, 20, 10, 3, 2, 1]));
```
Comment (wilzbach) [inline std/algorithm/sorting.d]: `make -f publictests` or [CircleCi](https://circleci.com/gh/dlang/phobos/1035?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)

```
out/std_algorithm_sorting.d(169): Error: no property 'retro' for type 'Merge!("a < b", short[], int[], double[])'
```

File: std/algorithm/sorting.d [hunk 33]
```diff
@@ -1048,152 +1048,302 @@ unittest
...
+    assert(equal(m, [20, 30]));
+    m.popBack();
+    assert(equal(m, [20]));
+    m.popFront();
+    assert(m.empty);
+}
+
 ///
 @safe pure nothrow unittest
 {
```
Comment (wilzbach) [inline std/algorithm/sorting.d]: It might be a personal preference, but I show the simpler unittest example as first example to the user.

Comment (nordlow) [inline std/algorithm/sorting.d]: Done. Moved the longer unittest last.

File: std/algorithm/sorting.d (hunk 32 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: Done.

File: std/algorithm/sorting.d (hunk 31 above)
Comment (nordlow) [inline std/algorithm/sorting.d]: Done.

Comment (nordlow) [inline std/algorithm/sorting.d]: Don't you mean: cmake -f posix.mak publictests`?

Comment (wilzbach) [inline std/algorithm/sorting.d]: > Don't you mean: make -f posix.mak publictests?

Yes - sorry for the typo.

**Comment by nordlow** [comment]
Ready to pull?

File: std/algorithm/sorting.d [hunk 34]
```diff
@@ -1048,140 +1048,247 @@ unittest
...
-{
-private:
-    Rs _r;
-    alias comp = binaryFun!(less);
-    uint _crt;
+    private alias comp = binaryFun!less;
+    private alias ElementType = CommonType!(staticMap!(.ElementType, Rs));
+    private enum isBidirectional = allSatisfy!(isBidirectionalRange, staticMap!(Unqual, Rs));
+
+    debug private enum canCheckSortedness = isCopyable!ElementType && !hasAliasing!ElementType;
```
Comment (andralex) [inline std/algorithm/sorting.d]: Guess this would be a version:

```d
debug static if (isCopyable!ElementType && !hasAliasing!ElementType)
    version = can_check_sortedness;
```

Unsure on which is the best way.

**Comment by andralex** [comment]
Good engineering, thanks.
